nameko/testing/services.py

```diff
@@ -218,29 +218,62 @@ def cm_to_inches(self, cms):
 
 
 class MockDependencyProvider(DependencyProvider):
-    def __init__(self, attr_name):
+    def __init__(self, attr_name, dependency=None):
         self.attr_name = attr_name
-        self.dependency = MagicMock()
+        self.dependency = MagicMock() if dependency is None else dependency
 
     def get_dependency(self, worker_ctx):
         return self.dependency
 
 
-def replace_dependencies(container, *dependencies):
+def _replace_dependencies(container, **dependency_map):
+    if container.started:
+        raise RuntimeError('You must replace dependencies before the '
+                           'container is started.')
+
+    dependency_names = {dep.attr_name for dep in container.dependencies}
+
+    missing = set(dependency_map) - dependency_names
+    if missing:
+        raise ExtensionNotFound("Dependency(s) '{}' not found on {}.".format(
+            missing, container))
+
+    existing_providers = {dep.attr_name: dep for dep in container.dependencies
+                          if dep.attr_name in dependency_map}
+
+    for name, replacement in dependency_map.items():
+        existing_provider = existing_providers[name]
+        replacement_provider = MockDependencyProvider(
+            name, dependency=replacement)
+        container.dependencies.remove(existing_provider)
+        container.dependencies.add(replacement_provider)
+
+
+def replace_dependencies(container, *dependencies, **dependency_map):
     """ Replace the dependency providers on ``container`` with
-    :class:`MockDependencyProvider` objects if they are named in
-    ``dependencies``.
+    instances of :class:`MockDependencyProvider`.
 
-    Return the :attr:`MockDependencyProvider.dependency` of the replacements,
-    so that calls to the replaced dependencies can be inspected. Return a
-    single object if only one dependency was replaced, and a generator
-    yielding the replacements in the same order as ``names`` otherwise.
+    Any dependencies named in the `dependencies` args will be replaced with a
+    :class:`MockDependencyProvider` that provides a ``MagicMock()`` dependency.
+
+    The dependency provided by the ``MockDependencyProvider`` can be customised
+    by using the `dependency_map` kwargs. Each kwarg name is the dependency
+    name, and the value specifies the custom dependency that should be
+    provided.
```

Comment:
I would rewrite these two paragraphs as:

> Dependencies named in `*dependencies` will be replaced with a :class:`MockDependencyProvider`, which injects a `MagicMock` instead of the dependency.
> 
> Alternatively, you may use keyword arguments to name a dependency and provide the replacement value that the `MockDependencyProvider` should inject.